src/backend/metal/src/command.rs

```diff
@@ -1090,17 +1127,173 @@ impl com::RawCommandBuffer<Backend> for CommandBuffer {
+            let r = region.borrow();
+
+            // layer count must be equal in both subresources
+            debug_assert_eq!(r.src_subresource.layers.len(), r.dst_subresource.layers.len());
+            // aspect flags
+            // enforce equal formats of both textures
+            // TODO this should probably be "compatible" pixel formats instead of equal formats
+            debug_assert_eq!(src.raw.pixel_format(), dst.raw.pixel_format());
+            // enforce aspect flag restrictions
+            debug_assert_ne!(r.src_subresource.aspects.contains(Aspects::COLOR), r.src_subresource.aspects.contains(Aspects::DEPTH | Aspects::STENCIL));
+            debug_assert_ne!(r.dst_subresource.aspects.contains(Aspects::COLOR), r.dst_subresource.aspects.contains(Aspects::DEPTH | Aspects::STENCIL));
+            debug_assert_eq!(r.src_subresource.aspects, r.dst_subresource.aspects);
+            // check that we're only copying aspects actually in the image
+            debug_assert!(src.format.surface_desc().aspects.contains(r.src_subresource.aspects));
+
+            let only_one_depth_stencil = {
+                let has_depth = r.src_subresource.aspects.contains(Aspects::DEPTH);
+                let has_stencil = r.src_subresource.aspects.contains(Aspects::STENCIL);
+                has_depth ^ has_stencil
+            };
+
+            let src_size = range_size(&r.src_bounds);
+            let dst_size = range_size(&r.dst_bounds);
+            // In the case that the image format is a combined Depth / Stencil format,
+            // and we are only copying one of the aspects, we use the shader even if the regions
+            // are the same size
+            if src_size == dst_size && !(has_ds && only_one_depth_stencil) {
+                debug_assert!(is_offset_positive(&r.src_bounds.start));
+                debug_assert!(is_offset_positive(&r.dst_bounds.start));
+
+                blit_commands.push(soft::BlitCommand::CopyImage {
+                    src: src.raw.clone(),
+                    dst: src.raw.clone(),
+                    region: com::ImageCopy {
+                        src_subresource: r.src_subresource.clone(),
+                        src_offset: r.src_bounds.start,
+                        dst_subresource: r.dst_subresource.clone(),
+                        dst_offset: r.dst_bounds.start,
+                        extent: src_size,
+                    },
+                });
+            } else {
+                //TODO: fill up vertices
+            }
+        }
+
+        self.sink().blit_commands(blit_commands.into_iter());
+
+        if !blit_vertices.is_empty() {
+            let inner = unsafe { &mut *self.inner.get() };
+
+            // Note: we don't bother to restore any render states here, since we are currently
+            // outside of a render pass, and the state will be reset automatically once
+            // we enter the next pass.
+            let mut pipes = self.shared.service_pipes
```

Comment:
How do we enforce that we're outside of a renderpass (just curious)?

Reply:
I guess there is a bit of ambiguity on render passes: there are Vulkan RPs in the API, and Metal RPs in the implementation.
On the API side, Vulkan specification guarantees that `vkCmdBlitImage` is only valid outside of a render pass.
In the implementation, we can just stop an active render pass at any point, technically, and start a new one. But I think we aren't going to have an active one outside of an API RP anyway.

Reply:
Ok makes sense. So is there a reason we don't fail hard when `blit_image` is called inside the Vulkan/API RP within the Metal backend (since this can fail on the Vulkan backend)?